### PR TITLE
feat: support of secret providers in APIs

### PR DESCRIPTION
**Pull request by benoitgravitee**
## Issue

https://gravitee.atlassian.net/browse/APIM-7497

## Description

* Implement secret refs finder for API V4
* Call of new events related to secrets
* Adapt EL template variable provider processing
* include new secret-api dependency
* Add secrets-service plugin
* bump secret provider plugin version
* bump node version

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-gginyskqjw.chromatic.com)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -52,16 +52,17 @@
         <gravitee-cockpit-api.version>3.7.0</gravitee-cockpit-api.version>
         <gravitee-cloud-initializer.version>2.1.0</gravitee-cloud-initializer.version>
         <gravitee-spec-gen-api.version>1.1.0</gravitee-spec-gen-api.version>
-        <gravitee-common.version>4.5.1</gravitee-common.version>
+        <gravitee-common.version>4.6.0</gravitee-common.version>
+        <gravitee-secret-api.version>1.0.0-APIM-7497-api-secrets-SNAPSHOT</gravitee-secret-api.version>
```
Comment (graviteeio) [inline pom.xml:56]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:**  <gravitee-secret-api.version>************************************</gravitee-secret-ap  
[This comment was created by Aqua Pipeline]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml [hunk 2]
```diff
@@ -32,6 +32,10 @@
     <artifactId>gravitee-apim-rest-api-idp-core</artifactId>
     <name>Gravitee.io APIM - Management API - Identity Providers - Core</name>
 
+    <properties>
+        <gravitee-secret-api.version>1.0.0-APIM-7497-api-secrets-SNAPSHOT</gravitee-secret-api.version>
```
Comment (graviteeio) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml:36]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:**  <gravitee-secret-api.version>************************************</gravitee-secret-ap  
[This comment was created by Aqua Pipeline]

File: pom.xml (hunk 1 above)
Comment (graviteeio) [inline pom.xml:56]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:**  <gravitee-secret-api.version>************************************</gravitee-secret-ap  
[This comment was created by Aqua Pipeline]

File: pom.xml [hunk 3]
```diff
@@ -265,7 +266,7 @@
         <gravitee-resource-cache-redis.version>3.0.0-alpha.1</gravitee-resource-cache-redis.version>
         <gravitee-resource-oauth2-provider-keycloak.version>2.1.0</gravitee-resource-oauth2-provider-keycloak.version>
         <gravitee-service-geoip.version>3.0.0</gravitee-service-geoip.version>
-        <gravitee-secretprovider-kubernetes.version>1.0.1</gravitee-secretprovider-kubernetes.version>
+        <gravitee-secretprovider-kubernetes.version>2.0.0-APIM-7497-api-secrets-SNAPSHOT</gravitee-secretprovider-kubernetes.version>
```
Comment (graviteeio) [inline pom.xml:269]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:** etprovider-kubernetes.version>************************************</gravitee-secretpro  
[This comment was created by Aqua Pipeline]

File: pom.xml [hunk 4]
```diff
@@ -283,7 +284,7 @@
         <gravitee-reactor-message.version>5.0.0-alpha.11</gravitee-reactor-message.version>
         <gravitee-reactor-native-kafka.version>1.0.0-alpha.18</gravitee-reactor-native-kafka.version>
         <gravitee-apim-repository-bridge.version>5.1.2</gravitee-apim-repository-bridge.version>
-        <gravitee-secretprovider-hc-vault.version>1.0.0</gravitee-secretprovider-hc-vault.version>
+        <gravitee-secretprovider-hc-vault.version>2.0.0-APIM-7497-api-secrets-SNAPSHOT</gravitee-secretprovider-hc-vault.version>
```
Comment (graviteeio) [inline pom.xml:287]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:** cretprovider-hc-vault.version>************************************</gravitee-secretpro  
[This comment was created by Aqua Pipeline]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml (hunk 2 above)
Comment (graviteeio) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml:36]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:**  <gravitee-secret-api.version>************************************</gravitee-secret-ap  
[This comment was created by Aqua Pipeline]

File: pom.xml (hunk 3 above)
Comment (graviteeio) [inline pom.xml:269]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:** etprovider-kubernetes.version>************************************</gravitee-secretpro  
[This comment was created by Aqua Pipeline]

File: pom.xml (hunk 4 above)
Comment (graviteeio) [inline pom.xml:287]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:** cretprovider-hc-vault.version>************************************</gravitee-secretpro  
[This comment was created by Aqua Pipeline]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml (hunk 2 above)
Comment (graviteeio) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml:36]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:**  <gravitee-secret-api.version>************************************</gravitee-secret-ap  
[This comment was created by Aqua Pipeline]

File: pom.xml (hunk 4 above)
Comment (graviteeio) [inline pom.xml:287]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:** cretprovider-hc-vault.version>************************************</gravitee-secretpro  
[This comment was created by Aqua Pipeline]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml (hunk 2 above)
Comment (graviteeio) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml:36]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:**  <gravitee-secret-api.version>************************************</gravitee-secret-ap  
[This comment was created by Aqua Pipeline]

File: pom.xml (hunk 1 above)
Comment (graviteeio) [inline pom.xml:56]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:**  <gravitee-secret-api.version>************************************</gravitee-secret-ap  
[This comment was created by Aqua Pipeline]

File: pom.xml [hunk 5]
```diff
@@ -266,7 +267,7 @@
         <gravitee-resource-cache-redis.version>3.0.0-alpha.1</gravitee-resource-cache-redis.version>
         <gravitee-resource-oauth2-provider-keycloak.version>2.1.0</gravitee-resource-oauth2-provider-keycloak.version>
         <gravitee-service-geoip.version>3.0.0</gravitee-service-geoip.version>
-        <gravitee-secretprovider-kubernetes.version>1.0.1</gravitee-secretprovider-kubernetes.version>
+        <gravitee-secretprovider-kubernetes.version>2.0.0-APIM-7497-api-secrets-SNAPSHOT</gravitee-secretprovider-kubernetes.version>
```
Comment (graviteeio) [inline pom.xml:270]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:** etprovider-kubernetes.version>************************************</gravitee-secretpro  
[This comment was created by Aqua Pipeline]

File: pom.xml [hunk 6]
```diff
@@ -284,7 +285,7 @@
         <gravitee-reactor-message.version>5.0.0-alpha.11</gravitee-reactor-message.version>
         <gravitee-reactor-native-kafka.version>1.0.0-alpha.23</gravitee-reactor-native-kafka.version>
         <gravitee-apim-repository-bridge.version>5.1.2</gravitee-apim-repository-bridge.version>
-        <gravitee-secretprovider-hc-vault.version>1.0.0</gravitee-secretprovider-hc-vault.version>
+        <gravitee-secretprovider-hc-vault.version>2.0.0-APIM-7497-api-secrets-SNAPSHOT</gravitee-secretprovider-hc-vault.version>
```
Comment (graviteeio) [inline pom.xml:288]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:** cretprovider-hc-vault.version>************************************</gravitee-secretpro  
[This comment was created by Aqua Pipeline]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml (hunk 2 above)
Comment (graviteeio) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml:36]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:**  <gravitee-secret-api.version>************************************</gravitee-secret-ap  
[This comment was created by Aqua Pipeline]

Comment (graviteeio) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml:36]: ### :warning: Aqua detected sensitive data in your code  
**Category:** Config Generic   
**Description:** Generic Config Credential   
**Severity:** HIGH   
**Match:**  <gravitee-secret-api.version>************************************</gravitee-secret-ap  
[This comment was created by Aqua Pipeline]

**Review by guillaumelamirand — APPROVED** [review]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml (hunk 2 above)
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml]: override the version defined in the main pom.xml

File: gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml [hunk 7]
```diff
@@ -60,6 +64,11 @@
             <groupId>io.gravitee.node</groupId>
             <artifactId>gravitee-node-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.gravitee.secret</groupId>
+            <artifactId>gravitee-secret-api</artifactId>
+            <version>${gravitee-secret-api.version}</version>
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-idp/gravitee-apim-rest-api-idp-core/pom.xml]: the version should be defined in the dependency management of gravitee-apim-bom

File: gravitee-apim-integration-tests/pom.xml [hunk 8]
```diff
@@ -707,6 +707,11 @@
             <artifactId>gravitee-secret-provider-hc-vault</artifactId>
             <version>${gravitee-secretprovider-hc-vault.version}</version>
```
Comment (phiz71) [inline gravitee-apim-integration-tests/pom.xml:709]: should it be with scope `test`?

File: gravitee-apim-integration-tests/pom.xml [hunk 9]
```diff
@@ -707,6 +707,11 @@
...
         </dependency>
+        <dependency>
+            <groupId>com.graviteesource.service</groupId>
+            <artifactId>gravitee-service-secrets</artifactId>
+            <version>${gravitee-service-secrets.version}</version>
+        </dependency>
```
Comment (phiz71) [inline gravitee-apim-integration-tests/pom.xml:717]: should it be a with scope `test`?

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/pom.xml [hunk 10]
```diff
@@ -37,6 +37,11 @@
             <groupId>io.gravitee.gateway</groupId>
             <artifactId>gravitee-gateway-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.gravitee.secret</groupId>
+            <artifactId>gravitee-secret-api</artifactId>
+            <version>${gravitee-secret-api.version}</version>
```
Comment (phiz71) [inline gravitee-apim-gateway/gravitee-apim-gateway-reactor/pom.xml]: should be declard in the dependency managment of gravitee-apim-bom

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-sdk/pom.xml [hunk 11]
```diff
@@ -198,6 +198,12 @@
             <artifactId>gravitee-common</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>io.gravitee.secret</groupId>
+            <artifactId>gravitee-secret-api</artifactId>
+            <version>${gravitee-secret-api.version}</version>
```
Comment (phiz71) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-sdk/pom.xml]: should be declard in the dependency managment of gravitee-apim-bom

File: gravitee-apim-definition/gravitee-apim-definition-model/pom.xml [hunk 12]
```diff
@@ -47,6 +47,11 @@
             <groupId>io.gravitee.node</groupId>
             <artifactId>gravitee-node-vertx</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.gravitee.secret</groupId>
+            <artifactId>gravitee-secret-api</artifactId>
+            <version>${gravitee-secret-api.version}</version>
```
Comment (phiz71) [inline gravitee-apim-definition/gravitee-apim-definition-model/pom.xml]: should be declard in the dependency managment of gravitee-apim-bom

File: gravitee-apim-distribution/pom.xml [hunk 13]
```diff
@@ -1531,6 +1531,32 @@
...
+                    <version>${gravitee-service-secrets.version}</version>
+                    <type>zip</type>
+                    <scope>runtime</scope>
+                    <exclusions>
+                        <exclusion>
+                            <groupId>*</groupId>
+                            <artifactId>*</artifactId>
+                        </exclusion>
+                    </exclusions>
+                </dependency>
```
Comment (phiz71) [inline gravitee-apim-distribution/pom.xml]: Is it intended to add these 2 plugins in the default bundle?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-standalone/gravitee-apim-rest-api-standalone-distribution/src/main/assembly/plugin-assembly-distribution.xml [hunk 14]
```diff
@@ -169,6 +169,7 @@
                 <exclude>io.gravitee.policy:gravitee-gateway-services-ratelimit:zip</exclude>
                 <exclude>io.gravitee.reporter:*:zip</exclude>
                 <exclude>io.gravitee.tracer:*:zip</exclude>
+                <exclude>com.graviteesource.service:gravitee-service-secrets:zip</exclude>
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-standalone/gravitee-apim-rest-api-standalone-distribution/src/main/assembly/plugin-assembly-distribution.xml:172]: excluding a zip file here is not enough. Unfortunately, we still have 2 ways to build our release and our docker images.
The first one uses maven assembly plugin for our dev docker images.
The second one uses a script for our prod images.

You also need to change this script (https://github.com/gravitee-io/gravitee-api-management/blob/a7dd5bf221cfadad15a2aa1eb6c44f4829f93a2b/release/ci-steps/package-bundles.mjs#L156-L169) to add your exclusion

**Review by phiz71 — COMMENTED** [review]
One change for the release and multiple small changes for the dependencies.

File: gravitee-apim-definition/gravitee-apim-definition-model/pom.xml (hunk 12 above)
Comment (benoitgravitee) [reply gravitee-apim-definition/gravitee-apim-definition-model/pom.xml]: You are right it should be there, that is bad clean up

File: gravitee-apim-distribution/pom.xml (hunk 13 above)
Comment (benoitgravitee) [reply gravitee-apim-distribution/pom.xml]: The second for sure that is the new feature, considering the hc-vault (EE) is bundled, I assumed I might add this one.

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/pom.xml (hunk 10 above)
Comment (benoitgravitee) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/pom.xml]: will do

File: gravitee-apim-distribution/pom.xml (hunk 13 above)
Comment (benoitgravitee) [reply gravitee-apim-distribution/pom.xml]: added to the dev bundle

**Review by guillaumelamirand — APPROVED** [review]

**Pull request merged**